ocaml/networkd/bin/network_server.ml

```diff
@@ -27,11 +27,14 @@ let network_conf = ref "/etc/xcp/network.conf"
 
 let config : config_t ref = ref Network_config.empty_config
 
+let config_writable = ref true
```

Comment:
Adding this reference makes implementation more difficult (on sync_state it should be turned on, write, then turn off the switch)

I think calling write_config only when xapi calls `persist_state` makes sense

That means it should stopped being called on shutdown (`on_shutdown`) and periodically (`on_timer`) because it cannot be guaranteed at those points that the configuration is total

Reply:
I would have to see if at any time the configuration is changed other than this one point in xapi_pif, if so it may be worth keeping this.

Reply:
Taking a deeper look it does seem as if `Nm.bring_pif_up` is called in numerous places which is also the mechanism used to change the config. This mechanism using on_timer and on_shutdown seems to account for this change

Reply:
The problem is that on_timer and on_shutdown persist the configuration without ever ensuring that the configuration is not incomplete. This is why I think only the API for explicitely writing is the only way to write the configuration to disk. This allows xapi to ensure that the configuration is total and only then write it.

Reply:
The only place I can see this being an issue is in the on-startup iterator which clears the in-memory config and rebuilds it. I wouldn't expect to have an incomplete config in any other place. It is in this place, we need to disable writing so that even on those two calls the state is never persisted and remains to be the existing state. What other places in the code would there be an incomplete state?

Reply:
> What other places in the code would there be an incomplete state?

That's the thing, networkd has no way of knowing whether the configuration it holds is complete or not, this means it's open to write incomplete configurations when it writes by itself.

Example: xapi wants to send configuration for 8 interfaces
- xapi sends configuration for 4 interfaces
- networkd periodic timer ends, `on_timer` gets executed and networkd writes the configurations
- xapi sends configuration for 2 more interfaces
- xapi gets interrupted
- now you're left with an incomplete configuration on disk, precisely what we want to avoid in this PR

Reply:
> This is why config_writable is used to disable writing

The `sync_state` function is missing the disablement after having written the state, as I said on the first message. 

This error in the implementation is why I'm advising to go for a far simpler solution that is much harder to misuse.